### PR TITLE
feat(dialog): surface contentPadding prop to allow overriding default padding values dialog content FE-4758

### PR DESCRIPTION
**Pull request by edleeks87**
### Proposed behaviour

Surfaces `contentPadding` prop which has an interface that partially supports the styled-system
padding `p`, `py` and `px` props

### Current behaviour

No support for adjusting padding of `Dialog` content.

### Checklist

- [x] Commits follow our style guide
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

https://codesandbox.io/s/modest-forest-fzd81?file=/src/index.js

### CODE REVIEW

File: src/components/dialog/dialog.component.js [hunk 1]
```diff
@@ -34,12 +36,20 @@ const Dialog = ({
...
   const dialogRef = useRef();
   const innerContentRef = useRef();
   const titleRef = useRef();
   const listenersAdded = useRef(false);
+  const { p, py, px } = contentPadding;
+
+  const paddingTopInnerContent = [py, p].some(
+    (padding) => padding !== undefined
+  )
+    ? `${ZERO_PADDING}px`
```
Comment (grabkowski) [inline src/components/dialog/dialog.component.js:51]: I think that should be just 0 (without px), and we don't need to set it to a const as it's pretty self explanatory

File: src/components/dialog/dialog.component.js [hunk 2]
```diff
@@ -34,12 +36,20 @@ const Dialog = ({
...
+  contentPadding = {},
   ...rest
 }) => {
   const dialogRef = useRef();
   const innerContentRef = useRef();
   const titleRef = useRef();
   const listenersAdded = useRef(false);
+  const { p, py, px } = contentPadding;
+
+  const paddingTopInnerContent = [py, p].some(
```
Comment (grabkowski) [inline src/components/dialog/dialog.component.js:48]: could that be moved into the styled component?

File: src/components/dialog/dialog.component.js [hunk 3]
```diff
@@ -273,6 +289,12 @@ Dialog.propTypes = {
   bespokeFocusTrap: PropTypes.func,
   /** The ARIA role to be applied to the Dialog container */
   role: PropTypes.string,
+  /** Padding to be set on the Dialog content */
+  contentPadding: PropTypes.shape({
+    p: PropTypes.oneOf([0, 1, 2, 3, 4, 5, 6, 7, 8]),
```
Comment (samtjo) [inline src/components/dialog/dialog.component.js:294]: Did it come from UX to lock these props down to 8px multipliers only? As a user I would expect these to work in the same way as all our other padding props, where we can pass in any valid CSS string

File: src/components/dialog/dialog.config.js [hunk 4]
```diff
@@ -7,5 +7,7 @@ export const DIALOG_SIZES = [
   "large",
   "extra-large",
 ];
-
 export const TOP_MARGIN = 20;
+export const CONTENT_TOP_PADDING = 24;
```
Comment (samtjo) [inline src/components/dialog/dialog.config.js:11]: This was previously 20px if you look at `DialogInnerContentStyle`

Comment (edleeks87) [reply src/components/dialog/dialog.config.js:11]: It's a requirement of this ticket to update that value to 24

File: src/components/dialog/dialog.component.js (hunk 3 above)
Comment (edleeks87) [reply src/components/dialog/dialog.component.js:294]: Yeah I spoke with @harpalsingh about this and we agreed for now this basic interface is all that we'll need, if consumers request a more flexible one we can open it up at that point. The issue is the need to calculate the values for the form margins, if we allow all the variations (strings, px/ %/ em and so on) it becomes quite complex and messy to pull the values out.

File: src/components/dialog/dialog.config.js (hunk 4 above)
Comment (samtjo) [reply src/components/dialog/dialog.config.js:11]: Would be good to put all the intended changes in the PR description 👍🏻

File: src/components/dialog/dialog.style.js [hunk 5]
```diff
@@ -124,12 +164,13 @@ const DialogContentStyle = styled.div`
...
   flex: 1;
   padding: 0px ${HORIZONTAL_PADDING}px ${CONTENT_BOTTOM_PADDING}px;
+  ${paddingFn}
 `;
 
 const DialogInnerContentStyle = styled.div`
-  padding-top: 20px;
   position: relative;
   flex: 1;
+  ${paddingFn}
```
[the remark this answers is not in the document]
Comment (edleeks87) [reply src/components/dialog/dialog.style.js:173]: Yeah missed that one 🤦

**Review by grabkowski — DISMISSED** [review]

File: src/components/dialog/dialog.component.js (hunk 3 above)
Comment (samtjo) [reply src/components/dialog/dialog.component.js:280]: Ok makes sense 👍🏻

**Review by samtjo — APPROVED** [review]

**Review by divyajindel — APPROVED** [review]

**Review by grabkowski — DISMISSED** [review]

**Review by samtjo — DISMISSED** [review]

**Review by grabkowski — APPROVED** [review]

**Review by divyajindel — APPROVED** [review]

**Pull request merged**